### PR TITLE
Tests, CI, linting, and formatting

### PR DESCRIPTION
**Pull request by mortendahl**
Main changes:

- fix unit tests and CI
- switch to building on Python 3.6
- docker images are no longer built
- pylint has been removed (leaving only flake8)
- all code is now formatting using black

### CODE REVIEW

File: tf_encrypted/convert/convert.py [hunk 1]
```diff
@@ -351,7 +353,7 @@ def specop_from_numberedscope(scope):
 
 def strip_tensor_info(node_name: str) -> str:
   if node_name.startswith("^"):
-    return n[1:]
+    return node_name[1:]
```
Comment (mortendahl) [inline tf_encrypted/convert/convert.py]: @jvmncs does this look correct to you? was an on-the-fly bug fix

Comment (jvmncs) [reply tf_encrypted/convert/convert.py]: ya, I made this change as well!

File: tf_encrypted/convert/register.py [hunk 2]
```diff
@@ -138,11 +146,13 @@ def _conv2d(converter, node, inputs):
...
-                 shape,
-                 strides=int(max(node.attr["strides"].list.i)),
-                 padding=node.attr["padding"].s.decode('ascii'),
-                 channels_first=fmt == "NCHW")
+  layer = Conv2D(
+      x_in.shape.as_list(),
+      shape,
+      strides=int(max(node.attr["strides"].list.i)),
+      padding=node.attr["padding"].s.decode('ascii'),
+      channels_first=fmt == "NCHW"
```
Comment (jvmncs) [inline tf_encrypted/convert/register.py]: any reason why yapf isn't formatting these last lines in a standard way? sometimes it adds commas (which I think is preferable), but some like this one are missing one

File: tf_encrypted/convert/register.py [hunk 3]
```diff
@@ -308,13 +323,17 @@ def _strided_slice(converter, node: Any, inputs: List[str]) -> Any:
...
   strides = tf.constant(strides.attr["value"].tensor)
 
-  return tfe.strided_slice(input_out, begin, end,
-                                          strides=strides,
-                                          begin_mask=begin_mask,
-                                          end_mask=end_mask,
-                                          ellipsis_mask=ellipsis_mask,
-                                          new_axis_mask=new_axis_mask,
-                                          shrink_axis_mask=shrink_axis_mask)
+  return tfe.strided_slice(
```
Comment (jvmncs) [inline tf_encrypted/convert/register.py]: wow this is a huge improvement, not sure how this one was so bad before

File: tf_encrypted/keras/layers/convolutional.py [hunk 4]
```diff
@@ -438,16 +475,24 @@ def compute_output_shape(self, input_shape):
...
+    )
     for d in range(self.depth_multiplier):
       for i in range(in_channels):
         mask[:, :, i, i + (d * in_channels)] = 1.
     return np.transpose(mask, [0, 1, 3, 2])
 
   def set_weights(self, weights, sess=None):
-    """ Sets the weights of the layer.
+    """
+    Sets the weights of the layer.
```
Comment (jvmncs) [inline tf_encrypted/keras/layers/convolutional.py]: Is this recommended? I seem to recall Google docstrings having the first (imperative) sentence inline with the opening of the quotation

File: tf_encrypted/keras/layers/convolutional.py [hunk 5]
```diff
@@ -438,16 +475,24 @@ def compute_output_shape(self, input_shape):
     return [n_x, n_filters, h_out, w_out]
 
   def get_mask(self, in_channels):
-    mask = np.zeros((self.kernel_size[0],
-                     self.kernel_size[1],
-                     in_channels, in_channels * self.depth_multiplier))
+    """TODO"""
```
Comment (jvmncs) [inline tf_encrypted/keras/layers/convolutional.py]: if these new TODOs are common we should spin a "good first issue" out for them

File: tf_encrypted/keras/layers/normalization.py [hunk 6]
```diff
@@ -84,62 +85,83 @@ class BatchNormalization(Layer):
...
-    default_args_check(renorm_clipping,
-                       "renorm_clipping",
-                       "BatchNormalization")
-    default_args_check(virtual_batch_size,
-                       "virtual_batch_size",
-                       "BatchNormalization")
-    default_args_check(adjustment,
-                       "adjustment",
-                       "BatchNormalization")
+        moving_variance_initializer
```
Comment (jvmncs) [inline tf_encrypted/keras/layers/normalization.py]: another example of a missing comma

Comment (jvmncs) [reply tf_encrypted/keras/layers/normalization.py]: seems strange that it happens so sporadically

File: tf_encrypted/keras/engine/base_layer.py [hunk 7]
```diff
@@ -114,7 +116,7 @@ def add_weight(self, variable, make_private=True):
     return variable
 
   def set_weights(self, weights, sess=None):
-    """ Sets the weights of the layer.
+    """Sets the weights of the layer.
```
Comment (jvmncs) [inline tf_encrypted/keras/engine/base_layer.py]: According to Google style guide, this should be imperative "Set" and there should be a newline at the end of this line, I think?

File: setup.cfg [hunk 8]
```diff
@@ -6,16 +6,24 @@ markers =
...
-split_before_dict_set_generator=True
-coalesce_brackets=True
-# split_before_first_argument=True
+dedent_closing_brackets=True
+split_all_top_level_comma_separated_values=True
+split_before_first_argument=True
+split_penalty_after_opening_bracket=0
 
 [flake8]
 max-line-length=80
```
Comment (jvmncs) [inline setup.cfg]: I thought you were going to increase this? I'm cool with it like this, but just curious :)

File: tf_encrypted/keras/models/__init__.py [hunk 9]
```diff
@@ -5,9 +5,8 @@
 from .sequential import model_from_config
 from .sequential import clone_model
 
-
 __all__ = [
     'Sequential',
     'model_from_config',
-    'clone_model'
+    'clone_model',
```
Comment (jvmncs) [inline tf_encrypted/keras/models/__init__.py]: aaaaaaand here's an example where it was added??

File: tf_encrypted/convert/register.py (hunk 2 above)
Comment (mortendahl) [reply tf_encrypted/convert/register.py]: yapf doesn't add commas on its own, you have to manually do that, which in turn forces this formatting. however, it may choose to go with this formatting either way if the line will otherwise become too long. the Google style is less likely to do this but the Facebook style does it by default; we are somewhere in-between thanks in part to `dedent_closing_brackets=True`).

with black on the other hand, there doesn't seem to be a way to force it to use this formatting, and it may decide to keep everything on a single line after one line break, which is not ideal either imho.

File: tf_encrypted/convert/register.py (hunk 3 above)
Comment (mortendahl) [reply tf_encrypted/convert/register.py]: yeah, column alignment pure sucks :D

File: tf_encrypted/keras/layers/convolutional.py (hunk 4 above)
Comment (mortendahl) [reply tf_encrypted/keras/layers/convolutional.py]: yeah, docstrings need a do-over at some point; luckily flake8 seems to capture them (but currently ignored in `setup.cfg`)

File: tf_encrypted/keras/layers/normalization.py (hunk 6 above)
Comment (mortendahl) [reply tf_encrypted/keras/layers/normalization.py]: with yapf it has to be done by hand

File: tf_encrypted/keras/engine/base_layer.py (hunk 7 above)
Comment (mortendahl) [reply tf_encrypted/keras/engine/base_layer.py]: Believe you're right, let's fix in another issue

File: setup.cfg (hunk 8 above)
Comment (mortendahl) [reply setup.cfg]: think we should stick to standard as much as possible, and since we'll be using a formatter I think it's less of an issue

File: tf_encrypted/convert/register.py (hunk 2 above)
Comment (jvmncs) [reply tf_encrypted/convert/register.py]: yeah, that seems strange given black's stated goal of reducing diff churn. overall I'm okay with both if you are!

**Comment by mortendahl** [comment]
@jvmncs notice that I've now removed pylint! perhaps there's a way to salvage it, but at least the current cont didn't play nicely with black.

File: Makefile [hunk 10]
```diff
@@ -100,10 +89,19 @@ SPHINX_BUILD_GOOGLE_DOCSTRINGS = sphinx-apidoc
...
+
+BUILD_CONVERTER_README=$(CONVERT_DIR)/gen/readme_template.md
+$(BUILD_CONVERTER_README): $(BUILD_RESERVED_SCOPES) pythoncheck
+	python -m tf_encrypted.convert.gen.generate_reserved_scopes
+
 google-docstrings:
 	@$(SPHINX_BUILD_GOOGLE_DOCSTRINGS) -fMeET "$(SPHINX_PROJECT_DIR)" -o "$(SPHINX_NAPOLEAN_BUILD_DIR)"
 
-docs: google-docstrings
+docs: $(BUILD_CONVERTER_README) google-docstrings
```
Comment (jvmncs) [inline Makefile]: I know it's technically correct, but I don't really like the converter readme being built with the defunct Sphinx docs. this makes it really easy for people to forget to build it

File: examples/house-credit-default/main.py [hunk 11]
```diff
@@ -38,140 +46,153 @@
...
-  """Build a logistic regression model with tf.keras."""
-  model = keras.Sequential([
-      layers.Dense(1, use_bias=False, activation='sigmoid',
-                   input_shape=[input_shape]),
-  ])
+    """Build a logistic regression model with tf.keras."""
+    model = keras.Sequential(
+        [
+            layers.Dense(
+                1, use_bias=False, activation="sigmoid", input_shape=[input_shape]
```
Comment (jvmncs) [inline examples/house-credit-default/main.py]: it's times like these that black is a little excessive :D

File: examples/house-credit-default/main.py [hunk 12]
```diff
@@ -38,140 +46,153 @@
...
-  train_y = train_y_df[1:246005]
+    # first 80 percent for training
+    train_x = train_x_df[1:246005]
+    train_y = train_y_df[1:246005]
 
-  # other 20 percent for evaluating
-  eval_x = train_x_df[246006:len(train_x_df) - 1]
-  eval_y = train_y_df[246006:len(train_y_df) - 1]
+    # other 20 percent for evaluating
+    eval_x = train_x_df[246006 : len(train_x_df) - 1]
```
Comment (jvmncs) [inline examples/house-credit-default/main.py]: spacing around a colon seems strange :/

File: examples/simple-average/run.py [hunk 13]
```diff
@@ -4,47 +4,60 @@
...
+        ),  # pylint: disable=unexpected-keyword-arg
+        provide_input(
+            player_name="inputter-2"
+        ),  # pylint: disable=unexpected-keyword-arg
+        provide_input(
+            player_name="inputter-3"
+        ),  # pylint: disable=unexpected-keyword-arg
+        provide_input(
+            player_name="inputter-4"
+        ),  # pylint: disable=unexpected-keyword-arg
```
Comment (jvmncs) [inline examples/simple-average/run.py]: we can remove these linter statements now, right?

File: examples/house-credit-default/main.py (hunk 11 above)
Comment (mortendahl) [reply examples/house-credit-default/main.py]: indeed :)

File: examples/house-credit-default/main.py (hunk 12 above)
Comment (mortendahl) [reply examples/house-credit-default/main.py]: i vaguely recall them having a discussion re this somewhere if you're interested

File: examples/simple-average/run.py (hunk 13 above)
Comment (mortendahl) [reply examples/simple-average/run.py]: yes!

File: Makefile (hunk 10 above)
Comment (mortendahl) [reply Makefile]: what if we let `build` depend on `docs`?

Comment (jvmncs) [reply Makefile]: that sounds good to me!

Comment (mortendahl) [reply Makefile]: Addressed in https://github.com/tf-encrypted/tf-encrypted/issues/779
